pkg/services/oauthserver/oasimpl/service.go

```diff
@@ -0,0 +1,504 @@
+
+	if registration.RedirectURI != nil {
+		client.RedirectURI = *registration.RedirectURI
+	}
+
+	var errGenCred error
+	client.ClientID, client.Secret, errGenCred = s.genCredentials()
+	if errGenCred != nil {
+		s.logger.Error("Error generating credentials", "client", client.LogID(), "error", errGenCred)
+		return nil, errGenCred
+	}
+
+	s.logger.Debug("Save service account")
+	saID, errSaveServiceAccount := s.saveServiceAccount(ctx, client.Name, client.ServiceAccountID, client.SelfPermissions)
+	if errSaveServiceAccount != nil {
+		return nil, errSaveServiceAccount
+	}
+	client.ServiceAccountID = saID
+
+	grantTypes := s.computeGrantTypes(registration.Self.Enabled, registration.Impersonation.Enabled)
+	client.GrantTypes = strings.Join(grantTypes, ",")
+
+	// Handle key options
+	s.logger.Debug("Handle key options")
+	keys, err := s.handleKeyOptions(ctx, registration.Key)
+	if err != nil {
+		s.logger.Error("Error handling key options", "client", client.LogID(), "error", err)
+		return nil, err
+	}
+	if keys != nil {
+		client.PublicPem = []byte(keys.PublicPem)
+	}
+	dto := client.ToDTO()
+	dto.KeyResult = keys
+
+	hashedSecret, err := bcrypt.GenerateFromPassword([]byte(client.Secret), bcrypt.DefaultCost)
+	if err != nil {
+		s.logger.Error("Error hashing secret", "client", client.LogID(), "error", err)
+		return nil, err
+	}
+	client.Secret = string(hashedSecret)
+
+	err = s.sqlstore.SaveExternalService(ctx, client)
+	if err != nil {
+		s.logger.Error("Error saving external service", "client", client.LogID(), "error", err)
+		return nil, err
+	}
+	s.logger.Debug("Registered", "client", client.LogID())
+	return dto, nil
+}
+
+// randString generates a a cryptographically secure random string of n bytes
+func (s *OAuth2ServiceImpl) randString(n int) (string, error) {
+	res := make([]byte, n)
+	if _, err := rand.Read(res); err != nil {
+		return "", err
+	}
+	return base64.RawURLEncoding.EncodeToString(res), nil
+}
+
+func (s *OAuth2ServiceImpl) genCredentials() (string, string, error) {
+	id, err := s.randString(20)
+	if err != nil {
+		return "", "", err
+	}
+	// client_secret must be at least 32 bytes long
+	secret, err := s.randString(32)
+	if err != nil {
+		return "", "", err
+	}
+	return id, secret, err
+}
+
+func (s *OAuth2ServiceImpl) computeGrantTypes(selfAccessEnabled, impersonationEnabled bool) []string {
+	grantTypes := []string{}
+
+	// If the client has permissions, it can use the client credentials grant type
+	if selfAccessEnabled {
+		grantTypes = append(grantTypes, string(fosite.GrantTypeClientCredentials))
+	}
+
+	// If the client has impersonate permissions, it can use the JWT bearer grant type
+	if impersonationEnabled {
+		grantTypes = append(grantTypes, string(fosite.GrantTypeJWTBearer))
+	}
+
+	return grantTypes
+}
+
+func (s *OAuth2ServiceImpl) handleKeyOptions(ctx context.Context, keyOption *oauthserver.KeyOption) (*oauthserver.KeyResult, error) {
+	if keyOption == nil {
+		return nil, fmt.Errorf("keyOption is nil")
+	}
```

Comment:
If impersonation is not needed, we don't need a key, we shouldn't error out I think

Reply:
I feel like "key" should be an option of impersonation
```json
{
  "name": "Test App3",
  "redirect_uri": "https://localhost:3000/a/test-app/",
  "self": {
    "enabled": true,
    "permissions": [
      {"action": "dashboards:read", "scope": "dashboards:*"}
    ]
  },
  "impersonation": {
    "enabled": true,
    "groups": true,
    "permissions": [
      {"action": "dashboards:read", "scope": "dashboards:*"},
      {"action": "dashboards:write", "scope": "dashboards:*"},
      {"action": "dashboards:read", "scope": "folders:*"},
      {"action": "dashboards:write", "scope": "folders:*"},
      {"action": "folders:read", "scope": "folders:*"},
      {"action": "folders:write", "scope": "folders:*"}
    ],
    "key": { "generate": true }
  },
}
```

Reply:
Tracked in an issue to handle in another PR

Reply:
tracked in an issue